File: day1-kotlin-syntax-crash-course/java/TypeInferenceExample.java
/**
 * Day 1 - Type Inference in Java vs Kotlin
 *
 * Java:
 *   - Before Java 10: EXPLICIT types everywhere
 *   - Java 10+: var for local variable type inference
 *   - NEVER works for class fields, method parameters, or return types
 *
 * Kotlin:
 *   - Type inference works EVERYWHERE
 *   - Local variables, properties, function return types
 *   - Smart casts after null checks and type checks
 */
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.HashMap;

public class TypeInferenceExample {

    // ═══════════════════════════════════════════════════════════════
    // 1. JAVA: EXPLICIT TYPES for class fields (always)
    // ═══════════════════════════════════════════════════════════════

    // Java: Class fields ALWAYS need explicit type (no var for fields!)
    private static String name = "Java";           // Must say "String"
    private static int count = 0;                  // Must say "int"
    private static List<String> list = new ArrayList<>();  // Must say "List<String>"
    private static Map<String, Integer> map = new HashMap<>();  // Must say "Map<String, Integer>"

    // ═══════════════════════════════════════════════════════════════
    // 2. JAVA 10+ var (LOCAL variables only)
    // ═══════════════════════════════════════════════════════════════

    public static void main(String[] args) {

        // Java 10+ var — type inferred from initializer
        var message = "Hello";              // String
        var number = 42;                    // int
        var decimal = 3.14;                 // double
        var flag = true;                    // boolean
        var list2 = List.of(1, 2, 3);       // List<Integer>
        var map2 = Map.of("a", 1, "b", 2);  // Map<String, Integer>

        // ❌ Java var CANNOT be used for:
        // var x;  // COMPILE ERROR: no initializer
        // var y = null;  // COMPILE ERROR: cannot infer type from null

        System.out.println("message: " + message);
        System.out.println("list2: " + list2);

        // ═══════════════════════════════════════════════════════════════
        // 3. JAVA: Smart casts? NOPE — need instanceof + explicit cast
        // ═══════════════════════════════════════════════════════════════

        Object obj = "Hello";

        // Java: instanceof check + separate cast (verbose!)
        if (obj instanceof String) {
            String str = (String) obj;  // Must cast explicitly
            System.out.println("String length: " + str.length());
        }

        // Java 16+: pattern matching for instanceof (better!)
        if (obj instanceof String str) {  // Pattern variable "str" created
            System.out.println("String length (Java 16): " + str.length());
        }

        // ═══════════════════════════════════════════════════════════════
        // 4. JAVA: Generic type inference with diamond <>
        // ═══════════════════════════════════════════════════════════════

        // Java 7+: Diamond operator <>
        var arrayList = new ArrayList<String>();   // Type inferred as ArrayList<String>
        var hashMap = new HashMap<String, Integer>();  // Type inferred

        // ═══════════════════════════════════════════════════════════════
        // 5. JAVA: Type inference for method return types? NOPE
        // ═══════════════════════════════════════════════════════════════

        // Java: Method return types MUST be explicit
        // No equivalent of Kotlin's "fun greet(): String = ..."

        // ═══════════════════════════════════════════════════════════════
        // 6. COMPARISON TABLE
        // ═══════════════════════════════════════════════════════════════
        //
        //   Feature                  Java 17        Kotlin
        //   ──────────────────────── ────────────── ──────────────────────
        //   Local var inference      var (Java 10+) val/var (always)
        //   Class field inference    ❌ Never        ✅ Always
        //   Method return type       ❌ Never        ✅ Yes (auto)
        //   Method param type        ❌ Never        ⚠️ Sometimes (lambdas)
        //   Smart casts              ⚠️ Java 16+    ✅ Automatic
        //   Generic inference        ✅ Diamond <>   ✅ Diamond <>
        //   Null type inference      ❌ Never        ✅ String? vs String
        //
        // ═══════════════════════════════════════════════════════════════

        // Example: Java var limitations
        // var x;           // ❌ Error: no initializer
        // var y = null;    // ❌ Error: cannot infer
        // var z = { 1, 2 }; // ❌ Error: ambiguous array
        // var a = (String) null; // ❌ Error: cannot infer

        // Workaround in Java: explicit type
        String y = null;  // OK with explicit type

        // ═══════════════════════════════════════════════════════════════
        // 7. KEY TAKEAWAYS
        // ═══════════════════════════════════════════════════════════════
        //
        // Java var is LIMITED:
        //   - Only local variables
        //   - Must have initializer
        //   - Cannot be null
        //   - Cannot be method params or return types
        //
        // Kotlin inference is FULL:
        //   - Works for local vars AND properties (class fields)
        //   - Works for function return types
        //   - Works with nullable types (val x: String? = null)
        //   - Smart casts eliminate explicit casting
        //
        // ═══════════════════════════════════════════════════════════════
    }

    // ═══════════════════════════════════════════════════════════════
    // 8. JAVA: NO type inference for return types
    // ═══════════════════════════════════════════════════════════════

    // Java: Must specify return type explicitly
    public static String greet(String name) {
        return "Hello, " + name;
    }

    // Java: Generic type inference in method calls
    public static <T> List<T> listOf(T... items) {
        return List.of(items);
    }
    // Call: List<String> list = listOf("a", "b");  // Type inferred
}

File: day1-kotlin-syntax-crash-course/kotlin/TypeInferenceExample.kt
/**
 * Day 1 - Type Inference in Kotlin
 *
 * Kotlin's type inference is MUCH more powerful than Java's:
 *   - Works for local variables, properties (class fields), and function return types
 *   - Smart casts after null checks and type checks
 *   - Generic type inference works everywhere
 */

// ═══════════════════════════════════════════════════════════════
// 1. TOP-LEVEL TYPE INFERENCE (Java CANNOT do this!)
// ═══════════════════════════════════════════════════════════════

// Kotlin: Top-level properties get type inference (Java: explicit type required)
val name = "Kotlin"                    // String
val count = 0                          // Int
val decimal = 3.14                     // Double
val isValid = true                     // Boolean
val list = listOf(1, 2, 3)            // List<Int>
val map = mapOf("a" to 1, "b" to 2)  // Map<String, Int>

// ═══════════════════════════════════════════════════════════════
// 2. CLASS PROPERTY TYPE INFERENCE (Java CANNOT do this!)
// ═══════════════════════════════════════════════════════════════

// Kotlin: Class properties get type inference
class User {
    val name = "Alice"       // Inferred as String (Java: private final String name = "Alice")
    var age = 25             // Inferred as Int (Java: private int age = 25)
    val scores = listOf(90, 85, 95)  // Inferred as List<Int>
}

// Java equivalent:
// class User {
//     private final String name = "Alice";   // Must specify String
//     private int age = 25;                  // Must specify int
//     private final List<Integer> scores = List.of(90, 85, 95);  // Must specify List<Integer>
// }

// ═══════════════════════════════════════════════════════════════
// 3. FUNCTION RETURN TYPE INFERENCE (Java CANNOT do this!)
// ═══════════════════════════════════════════════════════════════

// Kotlin: Return type inferred from expression body
fun add(a: Int, b: Int) = a + b   // Return type inferred as Int
fun greet(name: String) = "Hello, $name"  // Return type inferred as String
fun isPositive(n: Int) = n > 0     // Return type inferred as Boolean

// Java equivalent:
// public static int add(int a, int b) { return a + b; }  // Must say "int"
// public static String greet(String name) { return "Hello, " + name; }  // Must say "String"

// ═══════════════════════════════════════════════════════════════
// 4. SMART CASTS (Kotlin automatic after type checks)
// ═══════════════════════════════════════════════════════════════

fun processValue(value: Any): String {
    // Java 16+: if (value instanceof String s) { ... }
    // Kotlin: same but works with nullable, and more powerful

    return when (value) {
        is String -> "String of length ${value.length}"   // Smart cast: value is String
        is Int -> "Int: ${value * 2}"                     // Smart cast: value is Int
        is List<*> -> "List of size ${value.size}"        // Smart cast: value is List
        else -> "Unknown type"
    }
}

// Java equivalent:
// public static String processValue(Object value) {
//     if (value instanceof String s) {
//         return "String of length " + s.length();
//     } else if (value instanceof Integer i) {
//         return "Int: " + (i * 2);
//     } else if (value instanceof List<?> l) {
//         return "List of size " + l.size();
//     }
//     return "Unknown type";
// }

// ═══════════════════════════════════════════════════════════════
// 5. SMART CASTS after null checks
// ═══════════════════════════════════════════════════════════════

fun printLength(s: String?) {
    if (s != null) {
        // Smart cast: s is now String (not String?)
        println("Length: ${s.length}")  // No !! needed!
    }
}

// Java equivalent:
// public static void printLength(String s) {
//     if (s != null) {
//         System.out.println("Length: " + s.length());
//     }
// }

// ═══════════════════════════════════════════════════════════════
// 6. GENERIC TYPE INFERENCE
// ═══════════════════════════════════════════════════════════════

// Kotlin: Generic type inferred from arguments
val intList = listOf(1, 2, 3)          // List<Int>
val stringList = listOf("a", "b")      // List<String>
val mixedList = listOf(1, "two", 3.0)  // List<Any>

// Kotlin: Map type inference
val scores = mapOf("Alice" to 95, "Bob" to 87)  // Map<String, Int>

// Java: Also has diamond <> inference, but only for constructors
// var list = new ArrayList<String>();  // Java 10+

// ═══════════════════════════════════════════════════════════════
// 7. TYPE INFERENCE WITH LAMBDAS
// ═══════════════════════════════════════════════════════════════

// Kotlin: Lambda parameter types inferred
val numbers = listOf(1, 2, 3, 4, 5)
val evens = numbers.filter { it % 2 == 0 }  // "it" is inferred as Int

// Java: Also infers lambda params since Java 8
// numbers.stream().filter(n -> n % 2 == 0)

// ═══════════════════════════════════════════════════════════════
// 8. EXPLICIT TYPES (when needed)
// ═══════════════════════════════════════════════════════════════

// Sometimes you need explicit types (ambiguity, API design)
val explicitList: List<Int> = listOf(1, 2, 3)
fun explicitAdd(a: Int, b: Int): Int = a + b

// ═══════════════════════════════════════════════════════════════
// MAIN FUNCTION
// ═══════════════════════════════════════════════════════════════

fun main() {
    // Local variable type inference
    val message = "Hello"       // String
    val number = 42             // Int
    val decimal = 3.14          // Double
    val flag = true             // Boolean
    val list = listOf(1, 2, 3)  // List<Int>

    println("message: $message")
    println("number: $number")

    // Smart cast examples
    println(processValue("Hello"))   // String of length 5
    println(processValue(42))        // Int: 84
    println(processValue(listOf(1, 2))) // List of size 2

    // Null smart cast
    printLength("Kotlin")  // Length: 6
    printLength(null)      // (nothing printed)

    // ═══════════════════════════════════════════════════════════════
    // 9. COMPARISON SUMMARY
    // ═══════════════════════════════════════════════════════════════
    //
    // FEATURE                    KOTLIN              JAVA 17
    // ────────────────────────── ─────────────────── ───────────────────
    // Local variable inference   ✅ val/var x = ...   ✅ var x = ...
    // Property inference         ✅ val x = ...       ❌ Must specify type
    // Return type inference      ✅ fun f() = ...     ❌ Must specify type
    // Smart casts                ✅ Automatic         ⚠️ Java 16+ (limited)
    // Null type inference        ✅ String? vs String ❌ No null types
    // Lambda param inference     ✅ { it.length }     ✅ n -> n.length
    // Generic inference          ✅ listof(1,2)       ✅ diamond <>
    //
    // KEY TAKEAWAY:
    // Kotlin's type inference is UNIVERSAL — works everywhere
    // Java's var is LOCAL-ONLY — cannot be used for fields, returns, params
    //
    // ═══════════════════════════════════════════════════════════════
}
